### ISSUE TITLE
Replace embedded osgi packages by osgi-bundles from Maven-Central

### ISSUE THREAD

**Issue by HannesWell** [eclipse-equinox/equinox#18]
In M2E we would like to use `bnd.lib` 6.1 respectively 6.2 (see https://github.com/eclipse-m2e/m2e-core/pull/445) which imports newer version of the following OSGi packages than those that are embedded into `org.eclipse.osgi` respectively `org.eclipse.osgi.util`:

- `org.osgi.util.function` 1.2.0 (available 1.1.0)
- `org.osgi.util.promise` 1.2.0 (available 1.1.1)
- `org.osgi.util.tracker` 1.5.4 (available 1.5.3, which is actually sufficient because bnd.lib requires only 1.5)

Furthermore it requires the package the following package that is not embedded into any equinox bundle (at least I did not find it):
- `org.osgi.service.repository` 1.1.0

I assume it is desired to update those packages that are already available to the latest version?
Furthermore I assume to update them I have to copy the sources from the OSGi repository at the latest final release to the corresponding bundle in equinox:
https://github.com/osgi/osgi/releases/tag/r8-core-final-rerelease

How should we handle the missing `org.osgi.service.repository`?
M2E does not use the repository functionality (AFAICT) but it is required to resolve the bundle.
Should we fetch it from Maven-Central (and make it part of the M2E p2-repo and consequently contribute it to SimRel) or should it be embedded into an Equinox bundle too?

@tjwatson could you please provide some advice?

**Comment by HannesWell** [comment]
An update summary about the main PRs associated with this issue:
- https://github.com/eclipse-equinox/equinox.framework/pull/41
- https://github.com/eclipse-equinox/equinox/pull/30
- https://github.com/eclipse-equinox/equinox.bundles/pull/26
- https://github.com/eclipse-equinox/equinox.bundles/pull/39

Additionally there were multiple PRs to include the OSGi bundles into those features that contain the equinox-bundles that used to embed the replaced packages (if one is interested they are all referring this issue).

The only packages that could not have been replaced yet are the packages from `org.osgi.service.http` and `org.osgi.service.http.whiteboard` because they require a capability provided `org.apache.felix.http.servlet-api`. But its inclusion is currently blocked by https://github.com/eclipse-equinox/p2/pull/64.

**Comment by akurtakov** [comment]
This has caused javadoc generation failure. https://download.eclipse.org/eclipse/downloads/drops4/I20220618-1800/compilelogs/platform.doc.isv.javadoc.txt

**Comment by HannesWell** [comment]
> This has caused javadoc generation failure. https://download.eclipse.org/eclipse/downloads/drops4/I20220618-1800/compilelogs/platform.doc.isv.javadoc.txt

I'm about to prepare a PR that generalizes the pattern introduced in https://github.com/eclipse-platform/eclipse.platform.common/pull/19. I suspect this will fix the problem.

**Comment by akurtakov** [comment]
Isn't this one done now?

**Comment by HannesWell** [comment]
> Isn't this one done now?

Unfortunately not yet completely:

> The only packages that could not have been replaced yet are the packages from `org.osgi.service.http` and `org.osgi.service.http.whiteboard` because they require a capability provided `org.apache.felix.http.servlet-api`. But its inclusion is currently blocked by [eclipse-equinox/p2#64](https://github.com/eclipse-equinox/p2/pull/64).

But since we resolved #64, it should be possible to complete this after the december SimRel.
Then we should be able to announce the deprecation for removal of the bundles `org.eclipse.osgi.util` and `org.eclipse.osgi.services`. Maybe it would be reasonable to remove those packages from features as far as possible and only add them to the repo for those that really require them.

**Comment by tjwatson** [comment]
> But since we resolved #64, it should be possible to complete this after the december SimRel. Then we should be able to announce the deprecation for removal of the bundles `org.eclipse.osgi.util` and `org.eclipse.osgi.services`. Maybe it would be reasonable to remove those packages from features as far as possible and only add them to the repo for those that really require them.

Once we remove all the packages from the bundles and they only require-reexport the correct bundles, I would still not remove without adequate deprecation process to allow others to adjust.  Every time we have removed bundles like this it has caused much work for the community.  Here I'm not sure how much we gain doing so.  We do all the work to get the bundles to remain "working" with the solution of require-reexport only to then remove them quickly.

**Comment by laeubi** [comment]
But beside from products/feature no one should actually using the bundle directly? So probably its better to remove it and let people adjust their setup than retain it forever (in what case no one will ever adjust their builds regardless of deprecation).

**Comment by merks** [comment]
Keep in mind that Thomas speaks as someone with 20+ years of experience at a huge company where people scream when something breaks, while you speak as some who assumes that everyone in the world is eager, willing, and able to make adjustments, and, if not, too bad for them...

**Comment by laeubi** [comment]
Well if that "huge company" would pay "the contributors" to keep away the screaming people it might be worth to consider this.

Until that it puts the burden of maintenance (or not eager, unwilling, laziness or even dot care about their abuse) on the community that works for free... The OSGi spec is quite clear, that those bundles using this services have to import a certain **package** with a certain **version**, so anyone **requiring** the **bundle** does something that is very vendor specific, not portable and incur additional maintenance effort. So if this free "service" once vanished, yep its too bad for them, but probably better on the long run to adapt to a model that can easily catch up in the future than relying on bad practice for ever.

**Comment by merks** [comment]
The community that's affected is much closer to home:

![image](https://user-images.githubusercontent.com/208716/201672186-1990bbfb-8cdc-4181-b9fd-c4e54f941d6f.png)

These are not lazy people who are highly paid to accommodate every upstream improvement, but rather overworked people providing free goodness with limited time and budge.  Shifting burdens might seem like a great plan, depending where you are on the food chain, but you might find that if you over burden the food chain, only irrelevant microbes are left in the end...

**Comment by tjwatson** [comment]
> Well if that "huge company" would pay "the contributors" to keep away the screaming people it might be worth to consider this.
> 

What the hell does this mean?  I cannot even parse what actionable item you are trying to achieve with this statement.

My point is we have a deprecation policy, lets use it.  You want to break our consumers in the community "for their own good" is a hard sell.  What is also a hard sell is that we did all this effort to shell out the bundle, now we are close to completing that and the end result is to throw away the shelled out bundle?  Why did we do all this work in the first place!! We were shelling it out for a reason, not to break anyone and so we could source the classes from the true OSGi artifacts.  Now it has turned into some dig at the "huge companies" that are "holding us back", simply because I stated my desire not to break the community of plugins that live on top of this platform.

**Comment by laeubi** [comment]
> The community that's affected is much closer to home

Even with re-esport we at least need to include/adjust the new bundle there for this features... and everyone else not using such a feature already as well...

And as said, requiring the **bundle** because you want the **service** is just bad and binds you to the equinox bundle for no gain! And it also binds you unnecessarily to much more stuff as you really need, so this is not just an "upstream improvement" it is also an improvement fro the downstream as it increases interoperability, and  in the end even such dredged "we have duplicate providers" screams in the releng process, because we are more free to replace "bad" stuff.

And often the communities largely overlap, so well I started to fix this for PDE, so hopefully the "others" use-count can be decremented by one soon:
- https://github.com/eclipse-pde/eclipse.pde/pull/400

**Comment by tjwatson** [comment]
> And as said, requiring the **bundle** because you want the **service** is just bad and binds you to the equinox bundle for no gain!

Nobody is arguing that the require-bundle on the services and util bundles is a good practice.  I'm only stating it is a bad practice to quickly remove bundles when we do not fully understand who it will break.  I still get regular requests for help because we removed the `org.eclipse.equinox.ds` bundle long ago.  I'm not saying it was a bad idea that we removed that bundle, but the ripple effects are felt for a very long time.  And such removals inhibit something we all want, rapid adoption of new releases of the platform.

**Comment by laeubi** [comment]
> My point is we have a deprecation policy, lets use it.

I didn't disagreed with that, I just said that removing it (for whatever period the deprecation policy allow) will actually be a good thing! And even if we deprecate it people will still complain nerveless when this exact point in time will be unless we retain it forever.

> Now it has turned into some dig at the "huge companies" that are "holding us back", simply because I stated my desire not to break the community of plugins that live on top of this platform.

I'm sorry if that get in the wrong direction and never wanted to state that "someone is holding us back", I just replied to Ed mentioning that you care (probably more than me) much because you are working for a "huge company", so wanted to say that probably if some one would pay me for that I would also care more... but as it is not the case I'm not that concerned here and would be fine to remove it even right now.

**Comment by laeubi** [comment]
Also for platform:
- https://github.com/eclipse-platform/eclipse.platform/pull/268

**Comment by tjwatson** [comment]
I think we should look to add a the `deprecated` directive to our `Bundle-SymbolicName` headers now (ahead of any new R9 core specification).  And we can enhance tycho/PDE to report warnings when deprecated capabilities are used now.

**Comment by HannesWell** [comment]
Having a deprecated directive should make the remaining users more aware of the deprecation. I created https://github.com/eclipse-pde/eclipse.pde/issues/406 to add support in PDE for that.

And since Christoph said before actually that Plugin is not harmful and hopefully will be almost empty soon so that the maintenance effort for us will not be too great and should not worry much about a deprecation period. :)
By the way, how long should it be? The usual three years like for regular Java APIs?

**Comment by ffays** [comment]
At least upgrade `org.osgi.util.tracker` embedded packages from version `1.5.3` to version `1.5.4`.  

Even if the change is minor, there are more and more bundles requiring  
   `Import-Package: org.osgi.util.tracker; version="1.5.4"`  
E.g. `org.eclipse.jetty.osgi.boot` version `11.0.18` (which is the latest of Jetty 11 at the time of this post).  

Furthermore `org.osgi.util.tracker`  version `1.5.4` has been released on Oct 13, 2021 according to Maven Central, so two years ago ... so please please act!

**Comment by HannesWell** [comment]
> Furthermore org.osgi.util.tracker version 1.5.4 has been released on Oct 13, 2021 according to Maven Central, so two years ago ... so please please act!

This is actually not directly related to this issue, but it is fixed with https://github.com/eclipse-equinox/equinox/pull/394 and will be part of the December Release.

**Comment by ffays** [comment]
@HannesWell - Thank you for the follow-up

> This is actually not directly related to this issue, but it is fixed with #394 and will be part of the December Release.

To my point of view it is related to this issue, as the `org.eclipse.osgi` bundle is repacking the `org.osgi.util.tracker` package version 1.5.3, therefore it triggers an issue whenever there is a bundle having two separate dependency chains to package `org.osgi.util.tracker`, e.g. via `org.eclipse.osgi` bundle  version [3.18.500](https://mvnrepository.com/artifact/org.eclipse.platform/org.eclipse.osgi/3.18.500) exporting `org.osgi.util.tracker` package version 1.5.3  and `org.osgi.util.tracker` bundle version [1.5.4](https://mvnrepository.com/artifact/org.osgi/org.osgi.util.tracker/1.5.4) exporting `org.osgi.util.tracker` package version 1.5.4 .  

Quoting [BJ Hargrave](https://github.com/eclipse-equinox/equinox/issues/18#issuecomment-1157433262) comment hereby above in the thread:

>We (Eclipse collectively) should start using things from Maven Central instead of repackaging in things like org.eclipse.osgi.util (which is wrong). We really should phase this bundle out.

To which I totally agree.

Since this issue has been opened, the packages `org.osgi.util.function` and `org.osgi.util.promise` have been removed, but not `org.osgi.util.tracker` ... so why?

-/-

Bottom line: 

I'm glad to hear that the package `org.osgi.util.tracker` will be upgraded from 1.5.3 to 1.5.4 as part of the December `org.eclipse.osgi` bundle release.  

Therefore this upgrade will act as workaround for the time being.

**Comment by tjwatson** [comment]
> Since this issue has been opened, the packages `org.osgi.util.function` and `org.osgi.util.promise` have been removed, but not `org.osgi.util.tracker` ... so why?

Because this is part of the OSGi Core specification and must be exported by an OSGi R8 compliant framework.

**Comment by ffays** [comment]
> Because this is part of the OSGi Core specification and must be exported by an OSGi R8 compliant framework.

After checking, indeed the Tracker Specification are detailed in section 701, page 523 of the [OSGi Core specifications Release 8](https://docs.osgi.org/download/r8/osgi.core-8.0.0.pdf) documentation.

Thank you for mentioning it.

**Comment by HannesWell** [comment]
Once this is completed (probably via #403), all projects in SimRel has to be migrated off the o.e.osgi.service bundle as analyzed in
- https://github.com/eclipse-equinox/equinox/pull/403#issuecomment-1859081648.

**Comment by laeubi** [comment]
> Once this is completed (probably via https://github.com/eclipse-equinox/equinox/pull/403), all projects in SimRel has to be migrated off the o.e.osgi.service bundle

The bundle must/should not be used already now as it only exports API and never was meant to be required.

**Comment by HannesWell** [comment]
> > Once this is completed (probably via #403), all projects in SimRel has to be migrated off the o.e.osgi.service bundle
> 
> The bundle must/should not be used already now as it only exports API and never was meant to be required.

Absolutely.
But the entire message was mainly a reminder and should mention, that at least I will not take care of this before everything is done here. If anybody wants to migrate/reach out to all the affected SimRel projects before it is of course welcome. :)

**Comment by HannesWell** [comment]
Deprecation notice for `org.eclipse.osgi.services` added with:
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/1678

**Comment by HannesWell** [comment]
From https://github.com/eclipse-equinox/equinox/pull/403#issuecomment-1859081648

> Quite a few things use the to-be-deprecated bundle:
>
> ![image](https://github.com/eclipse-equinox/equinox/assets/208716/c91ab034-2729-48ab-8a11-5c9193215fe5)

I just created the the following PRs and issues to migrate all the dependencies in SimRel of o.e.osgi.services:
- https://github.com/eclipse-platform/eclipse.platform.ui/pull/1429
- https://github.com/eclipse/xtext/pull/2875
- https://github.com/eclipse-egit/egit/pull/8
- https://github.com/eclipse-rap/org.eclipse.rap/issues/160
- https://github.com/eclipse-cdt/cdt/issues/650
- https://github.com/eclipse-mylyn/org.eclipse.mylyn/issues/339
- https://github.com/eclipse-mpc/epp.mpc/issues/3
- https://github.com/eclipse-acceleo/acceleo/issues/127

I didn't found the webtools repository yet or are they still at bugzilla?

**Comment by tjwatson** [comment]
> I just created the the following PRs and issues to migrate all the dependencies in SimRel of o.e.osgi.services:

Thanks for the proactive contributions!  Just to make sure we are on the same page.  We still need to use the deprecation process so others outside of the Eclipse foundation projects have time to react.

**Comment by tjwatson** [comment]
Is there something in PDE that can post warnings when a deprecated bundle is referenced by a bundle project?

**Comment by laeubi** [comment]
> Is there something in PDE that can post warnings when a deprecated bundle is referenced by a bundle project?

`org.eclipse.pde.internal.core.builders.ManifestConsistencyChecker` can emit warnings / errors based on the manifest as far as I know but not sure anything is implemented yet.

**Comment by HannesWell** [comment]
> Just to make sure we are on the same page.  We still need to use the deprecation process so others outside of the Eclipse foundation projects have time to react.


Yes, we do. I just wanted to make sure we replace it sooner in the SimRel than later and some projects respond slowly any way😅
I already added a deprecation notice in the porting guide:
https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/1678

I could also add one to N&N


> Is there something in PDE that can post warnings when a deprecated bundle is referenced by a bundle project?

Not, yet but I have already created an issue a while ago:
https://github.com/eclipse-pde/eclipse.pde/issues/406

But it is not only about checking the manifest, also Features should be checked and the Plug-in project should be displayed differently e.g. striked through in the TP State view.

**Comment by HannesWell** [comment]
> Just to make sure we are on the same page.  We still need to use the deprecation process so others outside of the Eclipse foundation projects have time to react.


Yes, we do. I just wanted to make sure we replace it sooner in the SimRel than later and some projects respond slowly any way😅
I already added a deprecation notice in the porting guide:
https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/1678

I could also add one to N&N


> Is there something in PDE that can post warnings when a deprecated bundle is referenced by a bundle project?

Not, yet but I have already created an issue a while ago:
https://github.com/eclipse-pde/eclipse.pde/issues/406

But it is not only about checking the manifest, also Features should be checked and the Plug-in project should be displayed differently e.g. striked through in the TP State view.

**Comment by laeubi** [comment]
> Yes, we do. I just wanted to make sure we replace it sooner in the SimRel than later and some projects respond slowly😅

@merks now it is possible to deprecate bundles, should we have a SimRel rule to not include / reference deprecated bundles?

> I could also add one to N&N

If you do so please include this one as well:
- https://github.com/eclipse-equinox/equinox/pull/451

**Comment by merks** [comment]
Rules?!  I can bare get folks to read what I send them nor to get them to following existing rules.  😱  Adding new rules that they won't read and won't follow is unlikely to change this stark reality.  😖 

It would definitely be useful for the analyzer tool (*.aggran editor) to know that a bundle is deprecated so that I can determine who uses it and create issues to encourage those people to stop using them.  Is that information about deprecation available in the p2 metadata of the IU?  If not, can that be added as a property of the IU?   If I must to paw through actual artifacts, that is something that cannot be done quickly with an analysis tools that looks only at metadata...

**Comment by laeubi** [comment]
> Rules?! I can bare get folks to read what I send them nor to get them to following existing rules. 😱 Adding new rules that they won't read and won't follow is unlikely to change this stark reality. 😖

Without anything written down its just quite hard to argue why the project _must_ be changed, so its just an idea.

> Is that information about deprecation available in the p2 metadata of the IU?

I think I have seen manifest information in the p2 metadata so maybe one can check it from the ibuild repo if suitable information is there for the now deprecated `org.osgi.services` bundle but I'm not familiar with the `*.aggran editor` to judge.

**Comment by merks** [comment]
That information is not in the IU metadata:

![image](https://github.com/eclipse-equinox/equinox/assets/208716/2c20357d-df89-4667-8f50-ea94b90ae9f4)

The publisher action would need to add it.  If it were updated for that purpose, how long before there is a version of Tycho that uses it to build the 4.31 I-Builds?

**Comment by laeubi** [comment]
From OSGi POV I think its more to be expsected as part of the OSGI identity for example:
```
<provided namespace='osgi.identity' name='org.osgi.services' version='3.12'>
       <properties size='1'>
            <property name='type' value='osgi.bundle'/>
            <property name='deprecated' value='true'/>
        </properties>
</provided>
```
@tjwatson wdyt?

Beside that the touch points seem to include the symbolic name + version for a bundle.

**Comment by merks** [comment]
As long as it's in the p2 metadata somewhere I can make use of it...

**Comment by laeubi** [comment]
I think the place to change this might be `org.eclipse.osgi.internal.resolver.StateBuilder.createOsgiIdentityCapability(BundleDescription)` but not completely sure, this seems to be where BundlesAction is taking the metadata from the manifest.

**Comment by tjwatson** [comment]
> @tjwatson wdyt?

`osgi.identiy` capability is derived from the `Bundle-SymbolicName` header.  So attributes for the `osgi.identity` need to be specified there.  Note that capability attributes allowed by the OSGi specification are only 'String' | 'Version' | 'Long' | 'Double'

https://docs.osgi.org/specification/osgi.core/8.0.0/framework.module.html#d0e2870

So there is no Boolean to do something like:
```
Bundle-SymbolicName: org.eclipse.osgi.services; deprecated:Boolean="true"
```
You would have to use String (the default value type) "true" or "false" or use something else like Long where `0` is false and `1` is true.
```
Bundle-SymbolicName: org.eclipse.osgi.services; deprecated:Long="1"
```

> I think the place to change this might be `org.eclipse.osgi.internal.resolver.StateBuilder.createOsgiIdentityCapability(BundleDescription)` but not completely sure, this seems to be where BundlesAction is taking the metadata from the manifest.

Not sure what context you are using to reference that deprecated implementation of the old deprecated Equinox resolver API.  The framework does this here:

https://github.com/eclipse-equinox/equinox/blob/bb979606f8ed98d1641b5e76d62fe222c74fadb8/bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/container/builders/OSGiManifestBuilderFactory.java#L310-L316

**Comment by merks** [comment]
Note that this is already present:

https://github.com/eclipse-equinox/equinox/blob/bb979606f8ed98d1641b5e76d62fe222c74fadb8/bundles/org.eclipse.osgi.services/META-INF/MANIFEST.MF#L4

So I suppose we should rephrase the question to ask, how best to represent that?

**Comment by tjwatson** [comment]
> So I suppose we should rephrase the question to ask, how best to represent that?

Also note that is a directive (assignment with `:=`) vs an attribute (assignment with `=`).  I could go either way on if `deprecated` should be an attribute vs. a directive.  I've no idea if p2 represents anything about directives in its meta-data.

An attribute has an advantage that it can be used for matching or searching for capabilities in a repository, for example, by using a filter like `(deprecated=*)` to find all deprecated `osgi.identity`.

**Comment by laeubi** [comment]
> osgi.identiy capability is derived from the Bundle-SymbolicName header. So attributes for the osgi.identity need to be specified there. 

It is already there as mentioned by @merks 

>  Note that capability attributes allowed by the OSGi specification are only 'String' | 'Version' | 'Long' | 'Double'

Alright, it was just a quick idea, I think the best would be to just put in the message string, so given this example

https://github.com/eclipse-equinox/equinox/blob/bb979606f8ed98d1641b5e76d62fe222c74fadb8/bundles/org.eclipse.osgi.services/META-INF/MANIFEST.MF#L4

it should be 

```
<provided namespace='osgi.identity' name='org.osgi.services' version='3.12'>
       <properties size='1'>
            <property name='type' value='osgi.bundle'/>
            <property name='deprecated' value='Deprecated in favour of the original jars published by the OSGi working-group.'/>
        </properties>
</provided>
```

> Not sure what context you are using to reference that deprecated implementation of the old deprecated Equinox resolver API.

Sadly P2 is using the old API (or more formally some classes from it), if you look here (**public API!**):

https://github.com/eclipse-equinox/p2/blob/a795c7f4c05bc215cbe0048f833f7e1949daf412/bundles/org.eclipse.equinox.p2.publisher.eclipse/src/org/eclipse/equinox/p2/publisher/eclipse/BundlesAction.java#L181

`org.eclipse.osgi.service.resolver.BundleDescription` is used what only implementation seem to reside in the `org.eclipse.osgi.compatibility.state` it is a bit hard to track back where the capabilities are created but it leads me to the mentioned place.

If you see a way P2 can have its own implementation of `org.eclipse.osgi.service.resolver.BundleDescription`  (e.g. a `P2BundleDescription`  I think we can get rid of the compatibility stuff altogether in P2, the place to change would be `org.eclipse.equinox.p2.publisher.eclipse.BundlesAction.createBundleDescription(Dictionary<String, String>, File)` that seems the only place where `BundleDescription` is created.

**Comment by iloveeclipse** [comment]
Question from someone who didn't followed up the entire thread but sees hat `org.eclipse.osgi.services` bundle is now missing in the SDK - where it is this bundle is supposed to be now? I see at least Egit uses it, so "plain" SDK + Egit doesn't work anymore?

**Comment by iloveeclipse** [comment]
> So that should be fine, shouldn't it?

One needs also SDK + Egit **+ IBuild update site** to install Egit? One can't simply "drop in" something to SDK that needs org.eclipse.osgi.services? Would this bundle also on the final update site for the release?

Asking because now I can't simply "merge" SDK & Egit & the rest of my **offline** zip repositories in a script running **offline**, I need to download some extra bits before, and I don't understand **where** those bits are supposed to be coming from.

**Comment by laeubi** [comment]
The "bits" should come from the so called "target platform" (that is what egit is build against) so this target will/must include everything egit requires.

**Comment by iloveeclipse** [comment]
OK, the answer is also "we don't care".

**Comment by laeubi** [comment]
> OK, the answer is also "we don't care".

Don't understand what you wanted to say with that, I can't find out what egit uses as its prerequisites I can't even find a github repo (or do they use gitlab), but you seem to know it good enough to look there what they use as there target release.

**Comment by merks** [comment]
In all fairness, I don't think "we" understand what scripts and zips are involved in alternative ways to assemble an installation from parts.   So it's not an issue of not caring, but rather one of not understanding.  If the SDK installation doesn't need that bundle, then I'm not sure it's a strong argument to say, we should include it anyway because it makes some zip assembly scripts (that we don't understand) simpler...

**Comment by HannesWell** [comment]
> OK, the answer is also "we don't care".

Actually we do care, which is why I already reached out to all users and even provided PRs for those where I have the git repos cloned:
https://github.com/eclipse-equinox/equinox/issues/18#issuecomment-1861837389

For EGit especially there is already https://github.com/eclipse-egit/egit/pull/8, which was moved 
https://eclipse.gerrithub.io/c/eclipse-egit/egit/+/1173902
IIRC you are a commiter at EGit and I would appreciate a review/submission. :)

The reason why it is not in the SDK anymore is probably because the o.e.sdk feature does not (transitively) contain it anymore and nothing else in the SDK installation requires it so that it is pulled in (which I think is a good news).

As I have written here before or in https://github.com/eclipse-equinox/equinox/pull/403, o.e.osgi.services should stay in the eclipse/updates and simrel repo for the deprecation period so anything that requires it and has one of those p2-repos available can pull it in, if an installation is provisioned with P2.

**Comment by iloveeclipse** [comment]
> Actually we do care, which is why I already reached out to all users and even provided PRs

Thanks, I haven't seen that.

> IIRC you are a commiter at EGit and I would appreciate a review/submission. :)

I've not yet managed to migrate my account/setup to gerrithub, so I didn't saw the PR mail. I plan to migrate next year after vacation time.

**Comment by HannesWell** [comment]
No problem.

> > IIRC you are a commiter at EGit and I would appreciate a review/submission. :)
> 
> I've not yet managed to migrate my account/setup to gerrithub, so I didn't saw the PR mail. I plan to migrate next year after vacation time.

It is actually quite simple to setup. You can just go to the linked change, click on the login (through GitHub) button, enter your credentials and are then asked by GitHub to grant GerritHub (read) access to a few data from your Github account.

**Comment by HannesWell** [comment]
Just created the N&N entry for this:
- https://github.com/eclipse-platform/www.eclipse.org-eclipse/pull/124

**Comment by HannesWell** [comment]
Except for Acelleo nobody in SimRel uses o.e.osgi.serives anymore. And Accelleo also intends to replace it in time:
- https://github.com/eclipse-acceleo/acceleo/issues/127
